### ISSUE TITLE
API Proposal: Add PipeWriter accessor for the number of currently-buffered bytes  

### ISSUE THREAD

**Issue by roji** [dotnet/corefx#31074]
# The Problem

Applications use two major APIs when writing to a PipeWriter:

* `Advance()` "commits" bytes written to a previously-obtained Span or Memory, and
* `FlushAsync()` which makes all committed bytes visible to the reader.

If an applications needs to write a particularly large value (say a very large string), care must be taken not to buffer too much; doing so needless inflates memory use. In simple scenarios, an application should track how much it has already committed, and intermittently call `FlushAsync()` to provide a chance to the reader to read data, clearing out some buffers internally held by the pipe.

However, in some applications knowing how much has already been committed to the pipe isn't trivial: writing a single chunk may involve several components, which must currently coordinate between each other to make themselves aware of how much they've committed.

For a concrete example, when Npgsql (PostgreSQL database driver for .NET) sends a query to the database for execution, it sends multiple wire protocol messages (ideally packed in a single packet and flushed once, if they fit). One of these messages contains the list of parameters to the command, and one of these parameters can be an array of strings. The "type handler" responsible for writing the binary representation of a PostgreSQL array delegates to the string type handler for each of its values. Flushing may happen at any point while writing the array; it could be an array of one single huge string (so the flush would need to happen inside the string handler), or an array of many small strings (so the flush would need to happen inside the array handler). All these components would need to cooperate, passing a "currently buffered bytes" value back and forth.

# Proposed API

The idea would be to move the responsibility for tracking the number of committed/buffered bytes into PipeWriter, and expose it via a property:

```c#
class PipeWriter {
    public int BytesWritten { get; }
}

The number of bytes is o already present in the pipe (`_currentBytesWritten`), so it's mostly a matter of exposing it to users. It's expected that any other implementation of PipeWriter will similarly track buffered bytes.

Naming alternatives could be `BytesBuffered` (since `BytesWritten` may be taken to mean total bytes written in the history of the pipe).

/cc @davidfowl @mgravell @stephentoub

**Comment by pakrym** [comment]
It's unfortunate that adding a return type to `Advance` is a breaking change. Making return value of advance to be total number of buffered bytes would've been nice.

**Comment by davidfowl** [comment]
Right, I think the other interesting thing is that this ideally would go onto the `IBufferWriter<T>` interface as well.

**Comment by roji** [comment]
@benaadams it's true this can be added as a layer on top of `PipeWriter`. However, the main point of pipelines is to manage buffering when performing I/O, so it seems like a property exposing the amount buffered belongs in the class itself. Put another way, since pipelines is an abstraction - does it make sense to imagine an implementation of pipelines where a new `BytesWritten` property is irrelevant? If not then it probably belongs in pipelines.

@davidfowl I'm a bit less sure about `IBufferWrite<T>` - after all it's a much more abstract of idea of a sink, with no particular idea of buffering happening inside (or can the same really be said of `PipeWriter`?) With pipelines we can have a meaningful semantic definition of `BytesWritten`: as long as data are committed in the pipeline without yet being read (released) by the reader, that data is counted. With a general sink abstraction I'm much less sure.

**Comment by davidfowl** [comment]
I don't see why it's more relevant for the PipeWriter but not for the IBufferWriter. PipeWriter is IBufferWriter plus FlushAsync. The Pipe implementation the the only thing with the semantics you describe. PipeWriter is just an abstraction.

**Comment by svick** [comment]
Adding a method to an interface is a breaking change (at least until C# gets default interface methods https://github.com/dotnet/csharplang/issues/52). Doesn't that mean adding the method to `PipeWriter` is the only option here?

**Comment by roji** [comment]
@davidfowl ok, sorry, am still learning the new type hierarchy...

**Comment by davidfowl** [comment]
> Doesn't that mean adding the method to PipeWriter is the only option here?

It could mean that, it could also mean we do nothing and it's functionality built on top of a BufferWriter like abstraction.

**Comment by davidfowl** [comment]
We'll look at this in a future milestone
